### PR TITLE
Trial

### PR DESCRIPTION
**Pull request by Ivanner**
@A5hleyRich here's the pull request for the financial app trial

### CODE REVIEW

File: app/Http/Controllers/TransactionsController.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        if ($validator->fails()) {
+            return $validator->errors();
+        }
+        $transaction = new Transaction();
+        $transaction->label = $request->label;
+        $transaction->amount = $request->amount;
+        $transaction->date = $request->date;
+        $user->transactions()->save($transaction);
+        return $transaction->id;
+    }
```
Comment (A5hleyRich) [inline app/Http/Controllers/TransactionsController.php]: A Form Request would really clean this up.

https://laravel.com/docs/8.x/validation#form-request-validation

Is there any reason you're not using mass assignment? Then you could do:

```
$user->transactions()->save(new Transaction(
    $request->validated()
));
```

File: app/Http/Controllers/TransactionsController.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        }
+        /** @var User $user */
+        $user = $request->user();
+        /** @var Transaction $transaction */
+        $transaction = $user->transactions()->findOrFail($id);
+        $transaction->label = $request->label;
+        $transaction->amount = $request->amount;
+        $transaction->date = $request->date;
+        return $transaction->save();
+    }
```
Comment (A5hleyRich) [inline app/Http/Controllers/TransactionsController.php]: A Form Request and Model Binding would be preferred here.

https://laravel.com/docs/8.x/routing#route-model-binding

File: app/Http/Controllers/TransactionsController.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            // display errors on each form field
+            $errors = [];
+            foreach ($validator->errors()->getMessages() as $message) {
+                array_push($errors, $message[0]);
+            }
+            return ['message' => implode('. ', $errors)];
+        }
+
+        $file = $request->file('file');
+        $file->move(public_path(), $file->getClientOriginalName());
```
Comment (A5hleyRich) [inline app/Http/Controllers/TransactionsController.php]: The uploaded files will be publicly accessible. You should use `storage_path()` here.

File: app/Jobs/ImportProcessJob.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                if ($isHeader) {
+                    $isHeader = false;
+                    continue;
+                } else {
+                    $transaction = new Transaction();
+                    $transaction->user_id = $this->fileImport->user_id;
+                    $transaction->label = $row[0];
+                    $transaction->amount = floatval($row[1]);
+                    $transaction->date = $row[2];
+                    $transaction->save();
```
Comment (A5hleyRich) [inline app/Jobs/ImportProcessJob.php]: This won't fulfill this requirement:

> Imported entries should not appear in the balance list, until the entire import has completed.

**Review by A5hleyRich — COMMENTED** [review]
Thanks for taking the time to submit your trial project @Ivanner. I've reviewed this PR like I would any other member of the SpinupWP team so that you get a feel for our code review process. In addition to the comments, do you see any issue with the current import process? For example, if the import included 10,000 rows.

File: app/Http/Controllers/TransactionsController.php (hunk 1 above)
Comment (Ivanner) [reply app/Http/Controllers/TransactionsController.php]: Resolved.

File: app/Http/Controllers/TransactionsController.php (hunk 3 above)
Comment (Ivanner) [reply app/Http/Controllers/TransactionsController.php]: Resolved

File: app/Http/Controllers/TransactionsController.php (hunk 2 above)
Comment (Ivanner) [reply app/Http/Controllers/TransactionsController.php]: Resolved

File: app/Jobs/ImportProcessJob.php (hunk 4 above)
Comment (Ivanner) [reply app/Jobs/ImportProcessJob.php]: Temporarily set user_id to null and updated it when import completes.

**Comment by Ivanner** [comment]
> do you see any issue with the current import process?

@A5hleyRich  Yes, I updated it to use `createMany` so that all transactions are imported at once instead of one by one. Originally I had thought that a progress needed to be communicated to the user, but even if we need to do so, we can save multiple records in chunks and still achieve that goal.

**Pull request closed without merging**